### PR TITLE
Support pushing code

### PR DESCRIPTION
**Pull request by ankrgyl**
* Add support for `npx braintrust bundle`
* Move some typedefs around to support other changes
* Add errors, start, and progress messages to the stream
* Misc bug fixes and `any` conversions (as I ran into them)
* Support `messages` arg in invoke

### CODE REVIEW

**Review by manugoyal — COMMENTED** [review]
Looks good. Just one thought about code organization.

File: core/js/typespecs/functions.ts [hunk 1]
```diff
@@ -48,22 +53,41 @@ export const functionIdSchema = z
       .object({
         inline_context: runtimeContextSchema,
         code: z.string().describe("The inline code to execute"),
+        name: z
+          .string()
+          .optional()
```
[the remark this answers is not in the document]
Comment (ankrgyl) [reply core/js/typespecs/functions.ts:58]: I guess to be safe

File: core/js/typespecs/functions.ts [hunk 2]
```diff
@@ -48,22 +53,41 @@ export const functionIdSchema = z
...
+          .describe("The name of the inline code function"),
       })
       .describe("Inline code function"),
+    z
+      .object({
+        inline_prompt: promptDataSchema,
+        name: z.string().optional().describe("The name of the inline prompt"),
```
[the remark this answers is not in the document]
Comment (ankrgyl) [reply core/js/typespecs/functions.ts:65]: done

File: core/js/typespecs/functions.ts [hunk 3]
```diff
@@ -48,22 +53,41 @@ export const functionIdSchema = z
...
 export const invokeFunctionNonIdArgsSchema = z.object({
   input: z
     .any()
     .optional()
     .describe(
       "Argument to the function, which can be any JSON serializable value",
     ),
+  messages: z
+    .array(chatCompletionMessageParamSchema)
+    .optional()
```
[the remark this answers is not in the document]
Comment (ankrgyl) [reply core/js/typespecs/functions.ts:87]: invoke isn't stored so i don't think we need to do that

File: js/src/framework2.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
+import path from "path";
```
[the remark this answers is not in the document]
Comment (ankrgyl) [reply js/src/framework2.ts:1]: I'd prefer to leave as is for now -- this is just the next iteration of framework

File: js/src/framework2.ts [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+type NameOrId = { name: string } | { id: string };
+
+export type CreateProjectOpts = NameOrId;
+class ProjectBuilder {
+  create(opts: CreateProjectOpts) {
+    return new Project(opts);
+  }
+}
+export const project = new ProjectBuilder();
```
[the remark this answers is not in the document]
Comment (ankrgyl) [reply js/src/framework2.ts:19]: discussed offline -- will keep as is

**Review by manugoyal — APPROVED** [review]

**Pull request merged**